src/EditorFeatures/Core/Implementation/CommentSelection/ToggleBlockCommentCommandHandler.cs

```diff
@@ -0,0 +1,456 @@
+        {
+            DeleteText(textChanges, new TextSpan(spanToRemove.Start, commentInfo.BlockCommentStartString.Length));
+            var endMarkerPosition = spanToRemove.End - commentInfo.BlockCommentEndString.Length;
+            // Sometimes the block comment will be missing a close marker.
+            if (Equals(blockCommentSelection.GetSubstringFromText(endMarkerPosition, commentInfo.BlockCommentEndString.Length),
+                commentInfo.BlockCommentEndString))
+            {
+                DeleteText(textChanges, new TextSpan(endMarkerPosition, commentInfo.BlockCommentEndString.Length));
+            }
+        }
+
+        private class BlockCommentSelectionHelper
+        {
+            private readonly string _text;
+            private readonly ITextSnapshot _snapshot;
+
+            public TextSpan SelectedSpan { get; }
+
+            public IEnumerable<TextSpan> IntersectingBlockComments { get; }
+
+            public IEnumerable<TextSpan> UncommentedSpansInSelection { get; }
+
+            public BlockCommentSelectionHelper(IEnumerable<TextSpan> allBlockComments, SnapshotSpan selectedSnapshotSpan)
+            {
+                _text = selectedSnapshotSpan.GetText().Trim();
+                _snapshot = selectedSnapshotSpan.Snapshot;
+
+                SelectedSpan = TextSpan.FromBounds(selectedSnapshotSpan.Start, selectedSnapshotSpan.End);
+                IntersectingBlockComments = GetIntersectingBlockComments(allBlockComments, SelectedSpan);
+                UncommentedSpansInSelection = GetUncommentedSpansInSelection();
+            }
+
+            /// <summary>
+            /// Determines if the given span is entirely whitespace.
+            /// </summary>
+            public bool IsSpanWhitespace(TextSpan span)
+            {
+                for (var i = span.Start; i < span.End; i++)
+                {
+                    if (!IsLocationWhitespace(i))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+
+                bool IsLocationWhitespace(int location)
+                {
+                    var character = _snapshot.GetPoint(location).GetChar();
+                    return char.IsWhiteSpace(character);
+                }
+            }
+
+            /// <summary>
+            /// Determines if the location falls inside a commented span.
+            /// </summary>
+            public bool IsLocationCommented(int location)
+            {
+                return IntersectingBlockComments.Contains(span => span.Contains(location));
+            }
+
+            public bool DoesBeginWithBlockComment(CommentSelectionInfo commentInfo)
+            {
+                return _text.StartsWith(commentInfo.BlockCommentStartString, StringComparison.Ordinal)
+                       || _text.StartsWith(commentInfo.BlockCommentEndString, StringComparison.Ordinal);
```

Comment:
i don't undertstand the `||` part here... it doesn't make sense ot me.

Reply:
This checks if the selection begins with any comment marker.

Reply:
1.  That's odd (and needs explanation in code why that is desirable).
2. if that's the case, it should be renamed to StartsWithWithAnyBlockCommentMarker.  The 'begin' part really throws one off here as it sseems really strange to see if you're starting with a `*/` as opposed to a `/*`